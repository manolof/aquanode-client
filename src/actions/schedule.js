import { fetchSchedule, resetSchedule, setSchedule } from '../services/schedule';

export const ScheduleActions = Object.freeze({
	GET_START: '[Schedule] GET Start',
	GET_SUCCESS: '[Schedule] GET Success',
	GET_FAIL: '[Schedule] GET Fail',
	RESET_START: '[Schedule] RESET Start',
	RESET_SUCCESS: '[Schedule] RESET Success',
	RESET_FAIL: '[Schedule] RESET Fail',
	SET_START: '[Schedule] SET Start',
	SET_SUCCESS: '[Schedule] SET Success',
	SET_FAIL: '[Schedule] SET Fail',
});

export const getScheduleStartAction = () => ({
	type: ScheduleActions.GET_START,
});

export const getScheduleSuccessAction = (payload) => ({
	type: ScheduleActions.GET_SUCCESS,
	payload,
});

export const getScheduleFailAction = (error) => ({
	type: ScheduleActions.GET_FAIL,
	error,
});

export const getScheduleAction = () => (dispatch) => {
	dispatch(getScheduleStartAction());

	return fetchSchedule()
		.then((data) => dispatch(getScheduleSuccessAction(data)))
		.catch((err) => dispatch(getScheduleFailAction(err)));

};

export const resetScheduleStartAction = () => ({
	type: ScheduleActions.RESET_START,
});

export const resetScheduleSuccessAction = (payload) => ({
	type: ScheduleActions.RESET_SUCCESS,
	payload,
});

export const resetScheduleFailAction = (error) => ({
	type: ScheduleActions.RESET_FAIL,
	error,
});

export const resetScheduleAction = () => (dispatch) => {
	dispatch(resetScheduleStartAction());

	return resetSchedule()
		.then((data) => dispatch(resetScheduleSuccessAction(data)))
		.catch((err) => dispatch(resetScheduleFailAction(err)));

};

export const setScheduleStartAction = () => ({
	type: ScheduleActions.SET_START,
});

export const setScheduleSuccessAction = (payload) => ({
	type: ScheduleActions.SET_SUCCESS,
	payload,
});

export const setScheduleFailAction = (error) => ({
	type: ScheduleActions.SET_FAIL,
	error,
});

export const setScheduleAction = (value) => (dispatch) => {
	dispatch(setScheduleStartAction());

	return setSchedule(value)
		.then((data) => dispatch(setScheduleSuccessAction(data)))
		.catch((err) => dispatch(setScheduleFailAction(err)));

};
