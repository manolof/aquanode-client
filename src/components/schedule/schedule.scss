@import '../../styles/shared.scss';

.schedules {
	display: flex;
}

.schedule {
	display: flex;
	flex: 1;
	flex-direction: column;

	&__footer {
		display: flex;
		flex: 1;
		align-items: flex-end;
	}
}

button {
	&:not(:last-of-type) {
		margin-right: $spacing-l;
	}

	&.day {
		background-color: $color-warning;
	}

	&.night {
		background-color: $color-purple;
		color: $color-white;
	}
}
