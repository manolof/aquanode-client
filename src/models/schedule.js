import PropTypes from 'prop-types';

const Schedule = {
	override: PropTypes.bool,
	jobs: PropTypes.arrayOf(
		PropTypes.shape({
			job_state: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
			job_next_run: PropTypes.string,
		}),
	),
};

export const ScheduleModel = PropTypes.shape({
	lights: PropTypes.shape(Schedule),
	relay: PropTypes.shape(Schedule),
});
