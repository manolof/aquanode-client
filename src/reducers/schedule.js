import { ScheduleActions } from '../actions/schedule';

export const initialState = {
	override: false,
	items: [],
};

const scheduleReducer = (state = initialState, action) => {
	const items = action.payload;

	switch (action.type) {
		case ScheduleActions.GET_SUCCESS:
		case ScheduleActions.RESET_SUCCESS:
			return {
				...initialState,
				items,
			};

		case ScheduleActions.SET_SUCCESS:
			return {
				...initialState,
				override: true,
			};

		default:
			return state;
	}
};

export default scheduleReducer;
