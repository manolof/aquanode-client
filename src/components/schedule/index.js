import React, { Component } from 'react';

import { ScheduleModel } from '../../models/schedule';

class Schedule extends Component {
	render() {
		const { schedule } = this.props;

		return (
			<div className="Schedule">
				<h4>Schedule:</h4>

				{
					schedule.map((scheduleItem) =>
						<p key={scheduleItem.job_next_run}>{scheduleItem.job_name} : {scheduleItem.job_next_run}</p>,
					)
				}
			</div>
		);
	}
}

Schedule.propTypes = {
	schedule: ScheduleModel.isRequired,
};

export default Schedule;
