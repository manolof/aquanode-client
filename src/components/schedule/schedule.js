import PropTypes from 'prop-types';
import React, { Component } from 'react';
import moment from 'moment';

import './schedule.css';
import { ScheduleModel } from '../../models/schedule';

class Schedule extends Component {
	resetSchedule() {
		this.props.resetSchedule();
	}

	setSchedule(v) {
		this.props.setSchedule(v);
	}

	render() {
		const { schedule } = this.props;

		return (
			<div className="panel">
				<header className="panel__header">Schedule</header>

				<div className="panel__content">
					{
						schedule.map((scheduleItem) =>
							<p key={scheduleItem.job_next_run}>
								<strong>"{scheduleItem.job_name}"</strong> {moment(scheduleItem.job_next_run).format('MMM DD, HH:mm')}
							</p>,
						)
					}
				</div>

				<footer className="panel__footer">
					<button onClick={() => this.resetSchedule()}>RESET</button>
					<button className="day" onClick={() => this.setSchedule('day')}>SET DAY</button>
					<button className="night" onClick={() => this.setSchedule('night')}>SET NIGHT</button>
				</footer>
			</div>
		);
	}
}

Schedule.propTypes = {
	schedule: ScheduleModel.isRequired,
	resetSchedule: PropTypes.func.isRequired,
	setSchedule: PropTypes.func.isRequired,
};

export default Schedule;
