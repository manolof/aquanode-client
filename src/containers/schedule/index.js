import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { createGetScheduleAction } from '../../actions/schedule';
import Schedule from '../../components/schedule';
import { ScheduleModel } from '../../models/schedule';

class ScheduleContainer extends Component {
	componentDidMount() {
		const { fetchSchedule } = this.props;
		fetchSchedule();
	}

	render() {
		const { schedule } = this.props;

		return (
			<Schedule schedule={schedule}/>
		);
	}
}

ScheduleContainer.propTypes = {
	schedule: ScheduleModel.isRequired,
	fetchSchedule: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
	schedule: state.schedule,
});

const mapDispatchToProps = (dispatch) => ({
	fetchSchedule: () => dispatch(createGetScheduleAction()),
});

export { ScheduleContainer };

export default connect(mapStateToProps, mapDispatchToProps)(ScheduleContainer);
