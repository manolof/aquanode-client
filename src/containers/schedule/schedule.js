import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { createGetScheduleAction, resetScheduleAction, setScheduleAction } from '../../actions/schedule';
import Schedule from '../../components/schedule/schedule';
import { ScheduleModel } from '../../models/schedule';

class ScheduleContainer extends Component {
	componentDidMount() {
		const { fetchSchedule } = this.props;
		fetchSchedule();
	}

	render() {
		const { schedule, setSchedule, resetSchedule } = this.props;

		return (
			<Schedule schedule={schedule} resetSchedule={resetSchedule} setSchedule={setSchedule}/>
		);
	}
}

ScheduleContainer.propTypes = {
	schedule: ScheduleModel.isRequired,
	fetchSchedule: PropTypes.func.isRequired,
	setSchedule: PropTypes.func.isRequired,
	resetSchedule: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
	schedule: state.schedule,
});

const mapDispatchToProps = (dispatch) => ({
	fetchSchedule: () => dispatch(createGetScheduleAction()),
	setSchedule: (value) => dispatch(setScheduleAction(value)),
	resetSchedule: () => dispatch(resetScheduleAction()),
});

export { ScheduleContainer };

export default connect(mapStateToProps, mapDispatchToProps)(ScheduleContainer);
