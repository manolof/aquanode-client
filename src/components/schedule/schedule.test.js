import { shallow } from 'enzyme';
import React from 'react';

import { Schedule } from './schedule';

describe('Schedule Component', () => {
	let wrapper;
	const resetScheduleSpy = jest.fn();
	const setScheduleSpy = jest.fn();
	const mockScheduleProps = {
		override: false,
		items: [
			{
				job_name: 'night-17:0',
				job_next_run: '2018-08-23T15:00:00.000Z',
			},
			{
				job_name: 'day-7:0',
				job_next_run: '2018-08-23T07:00:00.000Z',
			},
		],
	};

	beforeEach(() => {
		wrapper = shallow(
			<Schedule schedule={mockScheduleProps} resetSchedule={resetScheduleSpy} setSchedule={setScheduleSpy}/>,
		);
	});

	afterEach(() => {
		jest.resetAllMocks();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('should render', () => {
		expect(wrapper.getElements()).toMatchSnapshot();
	});

});
