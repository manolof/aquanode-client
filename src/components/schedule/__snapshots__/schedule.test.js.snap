// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule Component should render 1`] = `
Array [
  <div
    className="panel"
  >
    <header
      className="panel__header"
    >
      <h3>
        Schedule
      </h3>
    </header>
    <div
      className="panel__content"
    >
      <main
        className="schedules"
      >
        <LightsSchedule
          resetSchedule={[Function]}
          schedule={
            Array [
              <p>
                <strong>
                  "day-5:0"
                </strong>
                 
                Aug 23, 07:00
              </p>,
              <p>
                <strong>
                  "day-7:0"
                </strong>
                 
                Aug 23, 09:00
              </p>,
            ]
          }
          setSchedule={[Function]}
        />
        <RelaySchedule
          resetSchedule={[Function]}
          schedule={
            Array [
              <p>
                <strong>
                  "day-7:0"
                </strong>
                 
                Aug 23, 09:00
              </p>,
              <p>
                <strong>
                  "day-9:0"
                </strong>
                 
                Aug 23, 11:00
              </p>,
            ]
          }
          setSchedule={[Function]}
        />
      </main>
    </div>
  </div>,
]
`;

exports[`Schedule Component should render in override mode 1`] = `
Array [
  <div
    className="panel"
  >
    <header
      className="panel__header"
    >
      <h3>
        Schedule
      </h3>
    </header>
    <div
      className="panel__content"
    >
      <main
        className="schedules"
      >
        <LightsSchedule
          resetSchedule={[Function]}
          schedule={
            <p>
              In override mode
            </p>
          }
          setSchedule={[Function]}
        />
        <RelaySchedule
          resetSchedule={[Function]}
          schedule={
            <p>
              In override mode
            </p>
          }
          setSchedule={[Function]}
        />
      </main>
    </div>
  </div>,
]
`;
